### ISSUE TITLE
Visualization tools

### ISSUE THREAD

**Issue by fchollet** [fchollet/keras#254]
I've started experimenting a little bit with experiment visualization. 

Here's a demo capable of monitoring a Keras experiment on a live-updating webpage: https://github.com/fchollet/hualos

What are your thoughts on what exactly we should be building? What do we need? 

Here are some initial ideas:
- typical workflow would require:
  - configure what you want to monitor and how via a callback constructor
  - be able to assign IDs to experiments and to monitor multiple experiments at the same time
- monitoring capabilities would include:
  - loss and validation loss monitoring
  - accuracy and validation accuracy monitoring _on a different scale_
  - confusion matrix monitoring for classification tasks
  - layers weights monitoring
  - activations monitoring

**Comment by fchollet** [comment]
Being able to interact with the experiment would also be a must. We should be building the Bloomberg terminal of Deep Learning.

**Comment by tleeuwenburg** [comment]
Sounds amazing, Francois. It's 8pm Sunday here. I'll be taking this for a
spin tomorrow during the day! Thanks so much for all your amazing work.

On 21 June 2015 at 12:48, François Chollet <notifications@github.com> wrote:

> I've started experimenting a little bit with experiment visualization.
>
> Here's a demo capable of monitoring a Keras experiment on a live-updating
> webpage: https://github.com/fchollet/hualos
>
> What are your thoughts on what exactly we should be building? What do we
> need?
>
> Here are some initial ideas:
>
>    -
>
>    typical workflow would require:
>     - configure what you want to monitor and how via a callback
>       constructor
>       - be able to assign IDs to experiments and to monitor multiple
>       experiments at the same time
>     -
>
>    monitoring capabilities would include:
>     - loss and validation loss monitoring
>       - accuracy and validation accuracy monitoring *on a different scale*
>       - confusion matrix monitoring for classification tasks
>       - layers weights monitoring
>       - activations monitoring
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fchollet/keras/issues/254>.
>



-- 
--------------------------------------------------
Tennessee Leeuwenburg
http://myownhat.blogspot.com/
"Don't believe everything you think"

**Comment by loyeamen** [comment]
The ability to show a graph of the train and validation loss / accuracy during the training phase would be quite helpful.

**Comment by aleju** [comment]
I have some code here that generates a plot for the loss and accuracy values (live updated during training). I could try to put that into a simple callback that can be added to `fit()` as a parameter, i.e. `fit(X, y, PlotterCallback())`. Example plot:
![plot](https://cloud.githubusercontent.com/assets/11698516/8271707/9c8e164a-1825-11e5-9769-4f3802cceecc.png)

**Comment by loyeamen** [comment]
If you could post the code that would be great
Thanks

**Comment by tleeuwenburg** [comment]
Hi,

So the first thing I found is that this doesn't work on Python 3.4 due to
gevent not being supported. Is Python 3.4 seen as a requirement? Apparently
there is a gevent fork which is compatible. I'll try to get a working
environment in a little while and look at the actual functionality.

Cheers,
-Tennessee

On 21 June 2015 at 12:48, François Chollet <notifications@github.com> wrote:

> I've started experimenting a little bit with experiment visualization.
>
> Here's a demo capable of monitoring a Keras experiment on a live-updating
> webpage: https://github.com/fchollet/hualos
>
> What are your thoughts on what exactly we should be building? What do we
> need?
>
> Here are some initial ideas:
>
>    -
>
>    typical workflow would require:
>     - configure what you want to monitor and how via a callback
>       constructor
>       - be able to assign IDs to experiments and to monitor multiple
>       experiments at the same time
>     -
>
>    monitoring capabilities would include:
>     - loss and validation loss monitoring
>       - accuracy and validation accuracy monitoring *on a different scale*
>       - confusion matrix monitoring for classification tasks
>       - layers weights monitoring
>       - activations monitoring
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fchollet/keras/issues/254>.
>



-- 
--------------------------------------------------
Tennessee Leeuwenburg
http://myownhat.blogspot.com/
"Don't believe everything you think"

**Comment by ganarajpr** [comment]
:+1: 
This is quite important. Could this perhaps be extended to be a GUI builder of NN's ?

**Comment by fchollet** [comment]
I do believe these tools should be framework-agnostic, ie. usable with different frameworks than Keras. Keras will interact with the visualization tools through a well-defined API, any framework that wants to plug into this API should be able to do so.

GUI NN building is not something I'm particularly interested in, since I believe text is more powerful than GUIs for manipulating complex systems. The main goal of these tools for the time being will be to expose everything that is going on inside an experiment, in a simple and visual way.

**Comment by fchollet** [comment]
> So the first thing I found is that this doesn't work on Python 3.4 due to
gevent not being supported. Is Python 3.4 seen as a requirement?

For now this is just a demo of how to get Keras and a webpage to communicate in real time, compatibility is not a concern. I'm sure we can replace gevent with something that will be Python 3 compatible when we move past the demo/exploration stage into actual implementation.

**Comment by donglixp** [comment]
I tried the demo. It is quite cool! It would be better if we plot the loss and accuracy in different figures. Maybe users can choose what they want to monitor (accuracy, loss, F1) in the dashboard. Another way is to enable users to define their own evaluation metric for the specific tasks, such as BLUE score for machine translation.

**Comment by tristandeleu** [comment]
I also just gave it a try and I really like it! I'm looking forward to move from console monitoring toward this centralized solution! In particular it would be very convenient to have the current batch (or training examples seen) per epoch and a progress bar like in `verbose=1`.

I also :+1: @donglixp regarding the metrics, even though I don't to what extent it would be possible.

**Comment by AntreasAntoniou** [comment]
Hey Francois,

Always a pleasure to use your libraries. Actually I would be very interested in a convolutional visualisation tool , so I can see what my network is "imagining" of you will. It's basically taking your load_weights method and mapping the tree of weights backwards towards the original image based on which pixel fires which neuron. Anyway awesome work man, my favourite deep learning lib right now next to Torch7 for lua

**Comment by bottydim** [comment]
Hey @fchollet ,

**Before anything you are awesome! Thank you for enabling me to graduate!** 

As part of my research, I have just set out to develop an activation visualization tool for keras RNN models. It would be great to contribute to the hualos project, but I believe a paramount necessity is a Remote Monitor type of a callback within the predict or even evaluate function. It would publish the current:

- input
- state values
- prediction
- target

What do you think?

**Comment by hadim** [comment]
To build a visualization tool, two things are needed : 

- A strategy implemented inside a Keras `Callback` to distribute the state of a `Model` during fitting.

   - Hualos strategy is to send information (json data) via a POST method. See the `RemoteMonitor` callback : https://github.com/fchollet/keras/blob/master/keras/callbacks.py#L369
    - TensorBoard strategy is to write all the information in a log folder that will be read by the visualization tool.

My point of view is that the POST method is great because it allows remote monitoring (as aetros does) but how do you handle the case when the fitting process has already started for a while and you decide to monitor ? The information is lost.

The log folder strategy allows you to access and explore 'Model' state when we want but the downside is that the folder need to be accessible for both the Keras instance and the visualization tool.

My opinion would go for the log folder strategy but I'd like to hear what do you think about that guys ?

- The other thing is the visualization tool it self. As TensorBoard and Hualos already do  I am in favour of a web based tool with a server. Flask/gevent combo not being very Python 3 friendly I would go with a library such as Tornado which is asynchronous and Python 3 friendly. But here again I would like to hear what do you guys think ?

The visualization tool should have the features highlighted by @fchollet in his first post of this issue.

**Comment by bottydim** [comment]
@hadim You are completely right!

Regarding the strategy, why don't we get the both of the two worlds i.e., publish the current information with a POST request and also keep a persistent object/file. We already have something similar in the form of the history object, which currently just stores the loss and accuracy. This way we could have both a **live-monitoring functionality** and **information availability** one. Additionally, the information could be explored after the experiment has concluded e.g., [check out LSTMVis](https://github.com/HendrikStrobelt/LSTMVis) . 

The idea is that the POST request will listen for the server response. If there have been any connection issues, the  `Callback`  would send the entire or only the missing parts from the log the moment the server becomes available again.

On the subject of the visualization tool, my thoughts were to build on top of Hualos and [LSTMViz](https://github.com/HendrikStrobelt/LSTMVis), but I agree with you that it should be a web-based application. Unfortunately, I have not worked with Tornado, but to my understanding, it is developed with '' long-lived connection to each user'' in mind, which is precisely what is going to happen with slowly trainable networks.

**Comment by wsstriving** [comment]
Hello, guys, When I use this tool, I always got an exception: 
` Warning: could not reach RemoteMonitor root server at 192.168.1.2:9000`

Details: I ran the python api.py on the machine with ip 192.168.1.2, and the server started normally, And then run the model training on the other machine (with gpu, but no monitor), by calling callback:
`remote = RemoteMonitor(root='192.168.1.2:9000')`

**Comment by MartinThoma** [comment]
It looks a bit as if fchollet/hualos is dead. @ollieglass did some improvements in https://github.com/ollieglass/hualos, as well as @Sebubu in https://github.com/Sebubu/hualos

@fchollet Do you want to keep developing Hualos or would you prefer if others continued the visualization project?

**Comment by bottydim** [comment]
@MartinThoma  I have a strong interest in visualizing and understanding neural networks - this is the topic of my PhD :D If you are keen I am happy to collaborate!

**Comment by fchollet** [comment]
I will no longer be developing Hualos, feel free to take over.

**Comment by shacharm2** [comment]
Does it still work with Keras ?
